lib/galaxy/objectstore/onedata.py

```diff
@@ -0,0 +1,557 @@
+        rel_path_dir = os.path.dirname(rel_path)
+        if not os.path.exists(self._get_cache_path(rel_path_dir)):
+            os.makedirs(self._get_cache_path(rel_path_dir), exist_ok=True)
+        # Now pull in the file
+        file_ok = self._download(rel_path)
+        self._fix_permissions(self._get_cache_path(rel_path_dir))
+        return file_ok
+
+    def _download(self, rel_path):
+        try:
+            dst_path = self._get_cache_path(rel_path)
+    
+            log.debug("Pulling file '%s' into cache to %s", rel_path, dst_path)
+
+            file_size = self._client.get_attributes(self.space_name, file_path=rel_path)['size']
+
+            # Test if cache is large enough to hold the new file
+            if not self.cache_target.fits_in_cache(file_size):
+                log.critical(
+                    "File %s is larger (%s) than the configured cache allows (%s). Cannot download.",
+                    rel_path,
+                    file_size,
+                    self.cache_target.log_description,
+                )
+                return False
+            
+            with open(dst_path, 'wb') as dst:
+                for chunk in self._client.iter_file_content(self.space_name, 
+                                                            STREAM_CHUNK_SIZE, 
+                                                            file_path=rel_path):
+                    dst.write(chunk)
+
+            log.debug("Pulled '%s' into cache to %s", rel_path, dst_path)
+
+            return True
+        except OnedataRESTError:
+            log.exception("Problem downloading file '%s'", rel_path)
+        return False
+
+    def _push_to_os(self, rel_path, source_file=None):
+        """
+        Push the file pointed to by ``rel_path`` to the object store under ``rel_path``. 
+        If ``source_file`` is provided, push that file instead while still using 
+        ``rel_path`` as the path.
+        """
+        try:
+            source_file = source_file if source_file else self._get_cache_path(rel_path)
+            if os.path.exists(source_file):
+                if os.path.getsize(source_file) == 0 and self._exists_in_onedata(rel_path):
+                    log.debug(
+                        "Wanted to push file '%s' to Onedata '%s' but its size is 0; skipping.", 
+                        source_file, 
+                        rel_path
+                    )
+                    return True
+                else:
+                    start_time = datetime.now()
+                    log.debug(
+                        "Pushing cache file '%s' of size %s bytes under '%s'",
+                        source_file,
+                        os.path.getsize(source_file),
+                        rel_path,
+                    )
+
+                    file_id = self._client.get_file_id(self.space_name, rel_path)
+
+                    with open(source_file, 'rb') as src:
+                        offset = 0
```

Comment:
po co ta zmienna skoro jej nie zmieniamy? czy nie brakuje czegos niżej tzn po kazdym chunku aktualizacja offsetu?

jeśli tak, to mają kijowe testy, chyba na małych plikach :P